docs: add sample usage of the FormArray API for reactive forms (#54138)
angular/angular: adev/src/content/guide/forms/reactive-forms.md

</docs-workflow>

## A more concrete usage of FormArray

> In a more concrete sample, we can imagine the following use case. You are application needs a basic form, with known information, but the user has the ability to add more or less sub-form.

Reviewer: ```suggestion
In a more concrete sample, we can imagine the following use case. Your application needs a basic form, with known information, but the user has the ability to add more or less sub-form.
```

Most of the time, these sub-forms are `FormGroup` parts of your principal form.

1. Define the `FormGroup`
2. Define the `FormArray`
3. Add action button to add entry
4. Add possibility to remove an entry

<docs-workflow>
<docs-step title="Define the FormGroup and template">

```ts
@Component({
  selector: 'my-form-comp',
  standalone: true,
  imports: [ReactiveFormsModule]

>   //component declaration

Reviewer: Do we need this comment?

Author: This comment is not revelant, I certainly forgot to remove it. If need I can remove it and make a new push.

Reviewer: Yes, remove anything that is not relevant.

</div>
```
At the start we have two defined controls : `controlOne` and `controlTwo`. We did not define the `FormArray` yet, we will do it on the next step.

</docs-step>

<docs-step title="Define the FormArray">

```ts
@Component({
  selector: 'my-form-comp',
  standalone: true,
  imports: [ReactiveFormsModule]
  //component declaration
})
export class FormArrayComp {
  formGroup : FormGroup = this.fb.group({
    controlOne: [''],
    controlTwo: [''],
    subGroups: this.fb.array<FormGroup>([])
  });
}
```

Here we have added the **subGroups** property to the `FormGroup` that contains the `FormArray`.

```html
 <div class="form-container">
  <form [formGroup]="formGroup">
    <input type="text" formControlName="controlOne" />
    <input type="text" formControlName="controlTwo" />
    
    @for(subForm of subForms.controls; subForm) {
      <div [formGroup]="subForm">
        <p>Sub-form {{$index + 1}}</p>
        <input type="text" formControlName="subControlName" />
      </div>
    }
    
    <button type="submit()">Submit</button>
  </form>
</div>
```

And in the template we have our principal form displayed and mapped, and a `@for` loop used to iterate over the `FormArray`.

> Tip: If you watch the HTML template, you will notice that the access to the `FormArray` is not from the `FormGroup` directly with the **get()** method. We will define a getter method.

Reviewer: ```suggestion
Tip: If you watch the HTML template, you will notice that the access to the `FormArray` is not from the `FormGroup` but directly using the **get()** method. We will define a getter method.
```

```ts
//adding a getter method
@Component({
  //boilerplate code
})
export class FormArrayComp {
  get subForms(): FormArray<FormGroup> {
    return this.form.get('subForms') as FormArray<FormGroup> ?? [];
  }
}
```
</docs-step>

<docs-step title="Add the action button to add entry">
Take back the HTML template and add a button to add an entry to the principal form : 

```html
<div class="form-container">
  <form [formGroup]="formGroup">
    <input type="text" formControlName="controlOne" />
    <input type="text" formControlName="controlTwo" />
    
    @for(subForm of subForms.controls; subForm) {
      <div [formGroup]="subForm">
        <p>Sub-form {{$index + 1}}</p>
        <input type="text" formControlName="subControlName" />
      </div>
    }
    
    <button type="submit()">Submit</button>
  </form>
  
  <button type="button" (click)="addEntry()">Add new entry</button>
</div>
```

Edit the component to add the method associated with the call to trigger with the click :

```ts
@Component({
  //boilerplate code
})
export class FormArrayComp {
  //boilerplate code
  addEntry() : void {
    this.form.push(this.fb.group({
      subControlName: ['']
    }));
  }
}
```

> With this method, when the user click on the button, it is adding a new `FormGroup` in the `FormArray`. When adding a new entry, the template will be updated.

Reviewer: ```suggestion
With this method, when the user clicks on the button, it is adding a new `FormGroup` in the `FormArray`. When adding a new entry, the template will be updated.
```

</docs-step>

<docs-step title="Add possibility to remove an entry">

> We can add entries to the `FormArray` but we should be able to remove a specific entry. The `FormArray` has the **removeAt()** method that permits to delete a specific index of the `FormArray`. 

Reviewer: ```suggestion
We can now add entries to the `FormArray` but we should also be able to remove a specific entry. The `FormArray` has the **removeAt()** method that permits to delete a specific index of the `FormArray`. 
```
